bindings/alicloud/sls/sls.go

```diff
@@ -0,0 +1,127 @@
+package sls
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"os/signal"
+	"strconv"
+	"syscall"
+	"time"
+
+	sls "github.com/aliyun/aliyun-log-go-sdk"
+	"github.com/aliyun/aliyun-log-go-sdk/producer"
+	"github.com/dapr/components-contrib/bindings"
+	"github.com/dapr/kit/logger"
+)
+
+type AliCloudSlsLogstorage struct {
+	logger   logger.Logger
+	producer *producer.Producer
+	metadata SlsLogstorageMetadata
+}
+
+type SlsLogstorageMetadata struct {
+	Endpoint        string `json:"endpoint"`
+	AccessKeyID     string `json:"accessKeyID"`
+	AccessKeySecret string `json:"accessKeySecret"`
+}
+
+// parse metadata field
+func (s *AliCloudSlsLogstorage) Init(metadata bindings.Metadata) error {
+	m, err := s.parseMeta(metadata)
+	if err != nil {
+		return err
+	}
+	s.metadata = *m
+	producerConfig := producer.GetDefaultProducerConfig()
+	// the config properties in the component yaml file
+	producerConfig.Endpoint = m.Endpoint
+	producerConfig.AccessKeyID = m.AccessKeyID
+	producerConfig.AccessKeySecret = m.AccessKeySecret
+	s.producer = producer.InitProducer(producerConfig)
+	return nil
+}
+
+func NewAliCloudSlsLogstorage(logger logger.Logger) *AliCloudSlsLogstorage {
+	logger.Debug("new a Slslog component object")
+	s := &AliCloudSlsLogstorage{
+		logger: logger,
+	}
+	return s
+}
+
+func (s *AliCloudSlsLogstorage) Invoke(ctx context.Context, req *bindings.InvokeRequest) (*bindings.InvokeResponse, error) {
+	// verify the metadata property
+	if logProject := req.Metadata["project"]; logProject == "" {
+		return nil, fmt.Errorf("SLS binding error: project property not supplied")
+	}
+	if logstore := req.Metadata["logstore"]; logstore == "" {
+		return nil, fmt.Errorf("SLS binding error: logstore property not supplied")
+	}
+	if topic := req.Metadata["topic"]; topic == "" {
+		return nil, fmt.Errorf("SLS binding error: topic property not supplied")
+	}
+	if source := req.Metadata["source"]; source == "" {
+		return nil, fmt.Errorf("SLS binding error: source property not supplied")
+	}
+
+	ch := make(chan os.Signal, 1)
+	signal.Notify(ch, syscall.SIGTERM, os.Interrupt)
+
+	// Start producer instancce
+	s.producer.Start()
```

Comment:
Should we take out Producer start in separate method to make `Invoke` look clean and crisp.

Reply:
Another thing to note is that this is working on `s.producer` which is a property of the object, and it's starting and stopping a producer in the same method.
Because there could be multiple calls to `Invoke` at the same time, this is not concurrency-safe

Reply:
I have fixed this problem